### ISSUE TITLE
error in pitch_area calculation

### ISSUE THREAD

**Issue by MannDoshi** [SRA-VJTI/Wall-E#63]
https://github.com/SRA-VJTI/Wall-E_v2.2-beta/blob/599e674feaee17bf6a6cf1048582e299eddff551/8_self_balancing/main/self_balancing.c#L53

![](https://plcgurus.net/wp-content/uploads/2018/03/classic_PID.png)

in line number 53 `pitch_error_cumulative` is multiplied with dt to get pitch_area but it should have been `pitch_area+=pitch_error*dt`. So why are we using multiplying `pitch_error_cumulative` with `dt`? 

TLDR;
We need to do summation (error*dt), but the current code does summation(error)*dt.

Submitted a possible fix in (#64)

**Comment by VedantParanjape** [comment]
@OSSome01 git blame shows you as the author. Please look into this.

CC: @laukik-hase @nilay994

**Comment by nilay994** [comment]
I don't understand the variables in the code, can you write a docstring next to variable declarations as to what they are holding?
For example:
```
static float prev_t = curr_t; /* holds the previous sample of elapsed time since start */
static float pitch_prev = pitch_curr; /* holds the pitch from the last iteration/sample */
static float pitch_err_cumulative = 0.0f; /* paired with Ki term, cumulative pitch error */
float dt = curr_t - prev_t;
float pitch_err = pitch_cmd - pitch_curr; /* pitch error */
pitch_err_cumulative += pitch_err * dt;
float pitch_rate = (pitch_curr - pitch_prev)/dt; /* rate of pitch, angular velocity body y axis */

prev_t = curr_t;
pitch_prev = pitch_curr;
```
also, from what it initially looks like, Ki term isn't doing anything, since `pitch_area` is not declared static.
https://github.com/SRA-VJTI/Wall-E_v2.2-beta/blob/5d6f57aca1bc935cafaf90dd7a430e2e6e3324b3/8_self_balancing/main/self_balancing.c#L36

Also, good job on 
```
	/**
	 * Integral term is bounded to prevent windup
	 * Bounding also helps, when dt is jumpy when the schedulers are warming up the threads
	*/
```
You could also start the integrator 5 seconds after bootup. The I gain here will help the bot make sure `pitch_curr` always becomes `pitch_cmd` irrespective of the CoM location of the bot (no steady state errors). Absence/less of Ki term could mean the bot tries to come close to `pitch_cmd`, but if you move the battery to the other side of CoM, the bot will now start going backwards, if it initially was drifting forwards.

**Comment by MannDoshi** [comment]
@nilay994 #56 will add proper commenting for the same.
Thanks for pointing out about pitch_area. It has been now declared static in #64

**Comment by VedantParanjape** [comment]
@nilay994 the comments are up, please check the latest version.

**Comment by nilay994** [comment]
> @nilay994 the comments are up, please check the latest version.

which PR should I review for that? And how close it this with pid-plotter-project? Is it being used to tune the gains after getting the realtime plots of setpoint vs current?

**Comment by VedantParanjape** [comment]
https://github.com/SRA-VJTI/Wall-E_v2.2-beta/pull/56 review this PR

No, we haven't yet used pid-plotter to tune this. We are tuning using a webpage, inshort, there's a rest-api to take input.

**Comment by saharshleo** [comment]
Testing this equations (the one with `dt`) is remaining
Use code from 093cce3 for testing them

**Comment by VedantParanjape** [comment]
> Testing these equations (the one with `dt`) is remaining
> Use code from [093cce3](https://github.com/SRA-VJTI/Wall-E_v2.2/commit/093cce392b746a703a2d403c995cfe66ae73526d) for testing them

Cool, post the results here.

**Comment by gautam-dev-maker** [comment]
This issue doesn't seem to be valid anymore.
